### DISCUSSION TITLE
Anyone know why we have `__private_attributes__` and `__class_vars__`???

### DISCUSSION THREAD

**Discussion by samuelcolvin** [pydantic/pydantic#4518] [Pydantic V2]
Do they really do different things, or can they be merged?

Someone in @pydantic/pydantic-maintainers might know?

**Comment by samuelcolvin** [comment]
@Bobronium and @layday might have some good ideas since they implemented `__private_attributes__` #1679 and `__class_vars__` #2179 respectively?

**Reply by Bobronium** [reply]
Without looking much into the implementation and based on what I remember and know about `__class_vars__` and `__private_attributes__`, first is to be used with model classes and second is to be used with model instances.

`__private_attributes__` were implemented to make it easier to declare and use instance attributes that not meant to be fields. They are stored `__slots__` opposed to fields that are stored in `__dict__`.

**Reply by layday** [reply]
It's been some time, but as I recall the purpose of `__class_vars__` was to reflect `ClassVar`s in subclasses, so that if a variable is annotated with `ClassVar` in a parent class and overridden in a subclass, it would not need to reannotated.

**Reply by samuelcolvin** [reply]
Thanks both.

In summary it looks like we need to keep both.
